File: packages/ride-recorder/android/src/main/java/sh/parkfi/riderecorder/RideRecorder.kt
package sh.parkfi.riderecorder

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.hardware.Sensor
import android.hardware.SensorEvent
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import android.os.Build
import android.os.Handler
import android.os.HandlerThread
import androidx.core.content.ContextCompat
import kotlin.math.max
import kotlin.math.sqrt

/**
 * Foreground IMU + barometer capture with a passive-detection state machine.
 * Android mirror of the iOS `RideRecorder`.
 *
 * Combines `TYPE_LINEAR_ACCELERATION` + `TYPE_GRAVITY` (→ specific force),
 * `TYPE_GYROSCOPE`, and `TYPE_PRESSURE` (→ relative altitude). Monitoring runs
 * at ~UI rate watching a rolling variance; a ride-start trigger escalates to
 * GAME rate and starts using the barometer. Battery is bounded by the JS layer
 * arming this only while in-park.
 */
class RideRecorder(context: Context) : SensorEventListener {
    var onRideStarted: (() -> Unit)? = null
    var onRideDetected: ((RideResult) -> Unit)? = null

    private val sensorManager = context.getSystemService(Context.SENSOR_SERVICE) as SensorManager
    private val linear = sensorManager.getDefaultSensor(Sensor.TYPE_LINEAR_ACCELERATION)
    private val gravity = sensorManager.getDefaultSensor(Sensor.TYPE_GRAVITY)
    private val gyro = sensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE)
    private val pressure = sensorManager.getDefaultSensor(Sensor.TYPE_PRESSURE)
    // TYPE_STEP_COUNTER is hardware-batched on a low-power hub; it keeps counting
    // through Doze / screen-off, so deltas read on the next foreground ping carry
    // the pocketed stretches too. Gated on ACTIVITY_RECOGNITION (API 29+) — see
    // stepsPermitted; without the grant we never register and stepSample() stays
    // null (the JS layer treats null as "no step sensor").
    private val stepCounter = sensorManager.getDefaultSensor(Sensor.TYPE_STEP_COUNTER)
    private val stepsPermitted =
        Build.VERSION.SDK_INT < Build.VERSION_CODES.Q ||
            ContextCompat.checkSelfPermission(context, Manifest.permission.ACTIVITY_RECOGNITION) ==
            PackageManager.PERMISSION_GRANTED

    private var thread: HandlerThread? = null
    private var handler: Handler? = null

    private var monitoring = false
    private var recording = false
    private var manual = false

    private val baroAvailable = pressure != null
    private val gyroAvailable = gyro != null

    // latest raw values, combined into a sample on each linear-accel event
    private val gravVec = DoubleArray(3) { if (it == 2) -9.81 else 0.0 }
    private val linVec = DoubleArray(3)
    private var gyroDegS = 0.0
    private var basePressure: Double? = null
    private var relAltitude: Double? = null

    // Session step counter (F-steps). TYPE_STEP_COUNTER reports a cumulative
    // since-boot value; the first event after arming becomes the baseline and
    // sessionSteps is the delta since then. `sessionStartMs` identifies the
    // session — the server keys its dedupe cursor on it, so it must change
    // exactly when the baseline resets. @Volatile: written on the sensor thread,
    // read from the plugin's call thread. Frozen (not cleared) on disarm,
    // mirroring iOS.
    @Volatile private var stepBaseline: Float? = null
    @Volatile private var sessionSteps: Int? = null
    @Volatile private var sessionStartMs: Long? = null

    data class StepSample(val steps: Int, val sessionStartMs: Long)

    private val varWindow = ArrayList<DoublePair>()
    private var highVarSince: Double? = null
    private var lowVarSince: Double? = null

    private val ring = ArrayList<RawSample>()
    private var capture = ArrayList<RawSample>()
    private var recordStart = 0.0

    private data class DoublePair(val t: Double, val v: Double)

    // IMU/baro sensors need no runtime grant; the step counter's
    // ACTIVITY_RECOGNITION gate is handled by the plugin (see RideRecorderPlugin).

    /** Cumulative steps since this monitoring session armed, with the session's
     *  start time as its identity; null when the device lacks a step counter,
     *  permission was denied, or never armed. */
    fun stepSample(): StepSample? {
        val steps = sessionSteps ?: return null
        val start = sessionStartMs ?: return null
        return StepSample(steps, start)
    }

    fun startMonitoring() {
        if (monitoring || linear == null) return
        monitoring = true
        stepBaseline = null
        val stepsSupported = stepCounter != null && stepsPermitted
        sessionSteps = if (stepsSupported) 0 else null
        sessionStartMs = if (stepsSupported) System.currentTimeMillis() else null
        thread = HandlerThread("ride-recorder").also { it.start() }
        handler = Handler(thread!!.looper)
        registerSensors(SensorManager.SENSOR_DELAY_UI)
    }

    fun stopMonitoring() {
        monitoring = false
        recording = false
        manual = false
        sensorManager.unregisterListener(this)
        thread?.quitSafely()
        thread = null
        handler = null
        ring.clear()
        capture.clear()
        varWindow.clear()
        highVarSince = null
        lowVarSince = null
        basePressure = null
    }

    fun startRecording() {
        if (!monitoring || recording) return
        manual = true
        beginRecording()
    }

    fun stopRecording(): RideResult? {
        if (!recording) return null
        return finishRecording()
    }

    private fun registerSensors(delay: Int) {
        sensorManager.unregisterListener(this)
        linear?.let { sensorManager.registerListener(this, it, delay, handler) }
        gravity?.let { sensorManager.registerListener(this, it, delay, handler) }
        gyro?.let { sensorManager.registerListener(this, it, delay, handler) }
        // pressure is inherently slow; SENSOR_DELAY_NORMAL is plenty
        pressure?.let { sensorManager.registerListener(this, it, SensorManager.SENSOR_DELAY_NORMAL, handler) }
        // Step counter: one event per step (or a hardware batch); NORMAL delay.
        // Re-registering across the UI↔GAME escalation is harmless — the value
        // is since-boot cumulative, so the baseline survives.
        if (stepsPermitted) {
            stepCounter?.let { sensorManager.registerListener(this, it, SensorManager.SENSOR_DELAY_NORMAL, handler) }
        }
    }

    override fun onAccuracyChanged(sensor: Sensor?, accuracy: Int) {}

    override fun onSensorChanged(event: SensorEvent) {
        when (event.sensor.type) {
            Sensor.TYPE_GRAVITY -> {
                gravVec[0] = event.values[0].toDouble()
                gravVec[1] = event.values[1].toDouble()
                gravVec[2] = event.values[2].toDouble()
            }
            Sensor.TYPE_GYROSCOPE -> {
                val x = event.values[0].toDouble()
                val y = event.values[1].toDouble()
                val z = event.values[2].toDouble()
                gyroDegS = sqrt(x * x + y * y + z * z) * 180 / Math.PI
            }
            Sensor.TYPE_STEP_COUNTER -> {
                val cum = event.values[0]
                // First event of the session (or a since-boot reset after a
                // reboot mid-session) establishes the baseline.
                val base = stepBaseline
                if (base == null || cum < base) stepBaseline = cum
                sessionSteps = (cum - (stepBaseline ?: cum)).toInt()
            }
            Sensor.TYPE_PRESSURE -> {
                val hpa = event.values[0].toDouble()
                val altM = SensorManager.getAltitude(
                    SensorManager.PRESSURE_STANDARD_ATMOSPHERE, hpa.toFloat()
                ).toDouble()
                if (basePressure == null) basePressure = altM
                relAltitude = altM - basePressure!!  // relative only — absolute is weather-biased
            }
            Sensor.TYPE_LINEAR_ACCELERATION -> {
                linVec[0] = event.values[0].toDouble()
                linVec[1] = event.values[1].toDouble()
                linVec[2] = event.values[2].toDouble()
                emitSample(event.timestamp / 1_000_000_000.0)  // ns → s
            }
        }
    }

    private fun emitSample(t: Double) {
        // specific force = linear acceleration + gravity (device frame), m/s²
        val sfx = linVec[0] + gravVec[0]
        val sfy = linVec[1] + gravVec[1]
        val sfz = linVec[2] + gravVec[2]
        val sfMs2 = sqrt(sfx * sfx + sfy * sfy + sfz * sfz)

        val gnorm = max(1e-6, sqrt(gravVec[0] * gravVec[0] + gravVec[1] * gravVec[1] + gravVec[2] * gravVec[2]))
        val sample = RawSample(
            t = t,
            sfMs2 = sfMs2,
            gx = gravVec[0] / gnorm,
            gy = gravVec[1] / gnorm,
            gz = gravVec[2] / gnorm,
            gyroDegS = gyroDegS,
            altRel = if (baroAvailable) relAltitude else null,
        )

        if (recording) {
            capture.add(sample)
            evaluateEnd(sample, t)
        } else {
            pushRing(sample, t)
            evaluateStart(sfMs2, t)
        }
    }

    private fun pushRing(s: RawSample, now: Double) {
        ring.add(s)
        while (ring.isNotEmpty() && now - ring.first().t > RideConst.RING_BUFFER_S) {
            ring.removeAt(0)
        }
    }

    private fun evaluateStart(sfMs2: Double, now: Double) {
        varWindow.add(DoublePair(now, sfMs2))
        while (varWindow.isNotEmpty() && now - varWindow.first().t > RideConst.VAR_WINDOW_S) {
            varWindow.removeAt(0)
        }
        if (varWindow.size <= 3) return
        val mean = varWindow.sumOf { it.v } / varWindow.size
        val variance = varWindow.sumOf { (it.v - mean) * (it.v - mean) } / varWindow.size
        if (variance > RideConst.START_VAR_THRESHOLD) {
            if (highVarSince == null) highVarSince = now
            else if (now - highVarSince!! >= RideConst.START_SUSTAIN_S) beginRecording()
        } else highVarSince = null
    }

    private fun evaluateEnd(s: RawSample, now: Double) {
        if (now - recordStart >= RideConst.MAX_DURATION_S) {
            finishRecording()
            return
        }
        if (manual) return
        val recent = capture.filter { now - it.t <= RideConst.VAR_WINDOW_S }
        if (recent.size <= 3) return
        val mean = recent.sumOf { it.sfMs2 } / recent.size
        val variance = recent.sumOf { (it.sfMs2 - mean) * (it.sfMs2 - mean) } / recent.size
        if (variance < RideConst.END_VAR_THRESHOLD) {
            if (lowVarSince == null) lowVarSince = now
            else if (now - lowVarSince!! >= RideConst.END_SUSTAIN_S) finishRecording()
        } else lowVarSince = null
    }

    private fun beginRecording() {
        if (recording) return
        recording = true
        highVarSince = null
        lowVarSince = null
        capture = ArrayList(ring)
        recordStart = capture.firstOrNull()?.t ?: ring.lastOrNull()?.t ?: 0.0
        registerSensors(SensorManager.SENSOR_DELAY_GAME) // escalate to ~50 Hz
        onRideStarted?.invoke()
    }

    private fun finishRecording(): RideResult? {
        if (!recording) return null
        recording = false
        val wasManual = manual
        manual = false
        val samples = capture
        capture = ArrayList()
        if (monitoring) registerSensors(SensorManager.SENSOR_DELAY_UI)

        val result = RideMetricsComputer.compute(samples, baroAvailable, gyroAvailable)
        if (result != null && !wasManual) onRideDetected?.invoke(result)
        return result
    }
}

File: packages/ride-recorder/ios/Sources/RideRecorderPlugin/RideRecorder.swift
import Foundation
import CoreMotion

/// Foreground IMU + barometer capture with a passive-detection state machine.
///
/// Monitoring runs `deviceMotion` at ~10 Hz (accel-only, cheap) and watches a
/// rolling 5 s variance of specific-force magnitude. When the ride-start trigger
/// fires it escalates to 50 Hz, starts the altimeter, seeds the capture with the
/// 10 s pre-trigger ring buffer, and emits `rideStarted`. On ride end (quiet for
/// 20 s or a 6 min cap) it computes metrics and emits `rideDetected`, then
/// reverts to monitoring. Callers arm/disarm this only while in-park (see the JS
/// `AchievementTracker`), which is what bounds battery.
final class RideRecorder {
    var onRideStarted: (() -> Void)?
    var onRideDetected: ((RideResult) -> Void)?

    private let motion = CMMotionManager()
    private let altimeter = CMAltimeter()
    private let pedometer = CMPedometer()
    private let queue = OperationQueue()
    // Background keep-alive so capture survives screen-lock (W9-B). Owned by the
    // monitoring lifecycle — started on arm, stopped on disarm.
    private let keepAlive = LocationKeepAlive()

    private var monitoring = false
    private var recording = false
    private var manual = false

    // Rolling window for the variance trigger (monitoring mode).
    private var varWindow: [(t: Double, v: Double)] = []
    private var highVarSince: Double? = nil
    private var lowVarSince: Double? = nil

    // 10 s pre-trigger ring buffer + the active capture.
    private var ring: [RawSample] = []
    private var capture: [RawSample] = []
    private var recordStart: Double = 0

    private var relAltitude: Double? = nil
    private var baroAvailable = false
    private var gyroAvailable = false

    // Session step counter (F-steps): cumulative steps since the current
    // monitoring session was armed, nil when step counting is unavailable or
    // never armed. `sessionStartMs` identifies the session — the server keys its
    // dedupe cursor on it, so it must change exactly when the counter resets.
    // Written on the pedometer's callback thread, read from the Capacitor call
    // thread — guarded by stepsLock. Frozen (not cleared) on disarm so the JS
    // layer's last read before/after park exit still resolves.
    private let stepsLock = NSLock()
    private var sessionStepsValue: Int? = nil
    private var sessionStartMsValue: Double? = nil
    // Dedicated instance for historical queries (reconciliation) — kept separate
    // from the live-updates instance to avoid interleaving update/query state.
    private let pedometerHistory = CMPedometer()

    // MARK: - Permissions

    static func permissionState() -> String {
        guard CMMotionActivityManager.isActivityAvailable() else { return "granted" }
        switch CMMotionActivityManager.authorizationStatus() {
        case .authorized: return "granted"
        case .denied, .restricted: return "denied"
        default: return "prompt"
        }
    }

    // MARK: - Lifecycle

    func startMonitoring(imuHz: Double, baroHz: Double) {
        guard motion.isDeviceMotionAvailable, !monitoring else { return }
        monitoring = true
        gyroAvailable = motion.isGyroAvailable
        queue.maxConcurrentOperationCount = 1
        // Hold a background location session so iOS keeps scheduling us with the
        // screen locked — otherwise deviceMotion updates stop within seconds of
        // suspension mid-ride (W9-B).
        keepAlive.start()
        startDeviceMotion(hz: imuHz)
        // Run the altimeter from arming, not from ride-start, so the 10 s
        // pre-trigger ring carries real altitude and the lift-hill climb is
        // captured (W3/F5). CMAltimeter is ~1 Hz — negligible battery vs the IMU.
        // relAltitude continuity across rides in one arming session is fine: all
        // metrics use relative deltas/drawdowns within a single capture.
        startAltimeter()
        // Steps ride the same arm/disarm lifecycle, which is what scopes them to
        // in-park time: the JS tracker arms only inside a geofence. CMPedometer
        // counts on the motion coprocessor, so backgrounded stretches (locked
        // phone in a pocket) are included when the next update lands.
        startPedometer()
    }

    func stopMonitoring() {
        monitoring = false
        recording = false
        manual = false
        keepAlive.stop()
        motion.stopDeviceMotionUpdates()
        altimeter.stopRelativeAltitudeUpdates()
        pedometer.stopUpdates()
        ring.removeAll()
        capture.removeAll()
        varWindow.removeAll()
        highVarSince = nil
        lowVarSince = nil
    }

    /// Cumulative steps since the current monitoring session armed, with the
    /// session's start time as its identity; nil when step counting is
    /// unavailable (no coprocessor, permission denied) or never armed. The
    /// server diffs successive cumulative reports against a per-session cursor.
    func stepSample() -> (steps: Int, sessionStartMs: Double)? {
        stepsLock.lock()
        defer { stepsLock.unlock() }
        guard let steps = sessionStepsValue, let start = sessionStartMsValue else { return nil }
        return (steps, start)
    }

    /// Historical step count over an absolute window (reconciliation). Served
    /// from the OS's ~7-day pedometer buffer, so it survives app kills and
    /// repairs anything the live session lost. Nil when unavailable.
    func queryStepSpan(fromMs: Double, toMs: Double, completion: @escaping (Int?) -> Void) {
        guard CMPedometer.isStepCountingAvailable(), toMs > fromMs else {
            completion(nil)
            return
        }
        let from = Date(timeIntervalSince1970: fromMs / 1000)
        let to = Date(timeIntervalSince1970: toMs / 1000)
        pedometerHistory.queryPedometerData(from: from, to: to) { data, _ in
            completion(data?.numberOfSteps.intValue)
        }
    }

    func startRecording() {
        guard monitoring, !recording else { return }
        manual = true
        beginRecording(seedFromRing: true)
    }

    func stopRecording() -> RideResult? {
        guard recording else { return nil }
        let result = finishRecording()
        return result
    }

    // MARK: - Capture

    private func startDeviceMotion(hz: Double) {
        motion.stopDeviceMotionUpdates()
        motion.deviceMotionUpdateInterval = 1.0 / hz
        motion.startDeviceMotionUpdates(using: .xArbitraryZVertical, to: queue) { [weak self] dm, _ in
            guard let self = self, let dm = dm else { return }
            self.handleMotion(dm)
        }
    }

    private func startAltimeter() {
        guard CMAltimeter.isRelativeAltitudeAvailable() else {
            baroAvailable = false
            return
        }
        baroAvailable = true
        relAltitude = nil
        altimeter.startRelativeAltitudeUpdates(to: queue) { [weak self] data, _ in
            guard let self = self, let data = data else { return }
            self.relAltitude = data.relativeAltitude.doubleValue
        }
    }

    private func startPedometer() {
        // isStepCountingAvailable() is true even when Motion & Fitness was
        // denied — gate on authorization too. Without this a denied device
        // freezes the sample at (0, start), every ping ships cum 0, and the
        // server's distance cap (creditedDistance) reads "moved but took zero
        // steps" — zeroing distance_m for that user forever. Denied ⇒ nil
        // sample ⇒ the server treats the device as pedometer-less and GPS
        // distance passes through unclamped.
        let auth = CMPedometer.authorizationStatus()
        guard CMPedometer.isStepCountingAvailable(), auth != .denied, auth != .restricted else {
            stepsLock.lock()
            sessionStepsValue = nil
            sessionStartMsValue = nil
            stepsLock.unlock()
            return
        }
        stepsLock.lock()
        sessionStepsValue = 0
        sessionStartMsValue = Date().timeIntervalSince1970 * 1000
        stepsLock.unlock()
        pedometer.startUpdates(from: Date()) { [weak self] data, error in
            guard let self = self else { return }
            guard let data = data else {
                // .notDetermined resolves at this first use — a denial at the
                // prompt lands here as an error. Nil the whole sample: a dead
                // pedometer must read as "absent", never as "0 steps".
                if error != nil {
                    self.stepsLock.lock()
                    self.sessionStepsValue = nil
                    self.sessionStartMsValue = nil
                    self.stepsLock.unlock()
                }
                return
            }
            self.stepsLock.lock()
            self.sessionStepsValue = data.numberOfSteps.intValue
            self.stepsLock.unlock()
        }
    }

    private func handleMotion(_ dm: CMDeviceMotion) {
        let t = dm.timestamp
        // Specific force = user acceleration + gravity, in G → magnitude in m/s².
        let sfx = dm.userAcceleration.x + dm.gravity.x
        let sfy = dm.userAcceleration.y + dm.gravity.y
        let sfz = dm.userAcceleration.z + dm.gravity.z
        let sfMs2 = (sfx * sfx + sfy * sfy + sfz * sfz).squareRoot() * RideConst.g

        let gyroDegS = (dm.rotationRate.x * dm.rotationRate.x
            + dm.rotationRate.y * dm.rotationRate.y
            + dm.rotationRate.z * dm.rotationRate.z).squareRoot() * 180 / .pi

        // gravity is already ~unit length in G; normalize defensively.
        let gnorm = max(1e-6, (dm.gravity.x * dm.gravity.x
            + dm.gravity.y * dm.gravity.y + dm.gravity.z * dm.gravity.z).squareRoot())

        let sample = RawSample(
            t: t,
            sfMs2: sfMs2,
            gx: dm.gravity.x / gnorm,
            gy: dm.gravity.y / gnorm,
            gz: dm.gravity.z / gnorm,
            gyroDegS: gyroDegS,
            altRel: baroAvailable ? relAltitude : nil
        )

        if recording {
            capture.append(sample)
            evaluateEnd(sample, now: t)
        } else {
            pushRing(sample, now: t)
            evaluateStart(sfMs2: sfMs2, now: t)
        }
    }

    private func pushRing(_ s: RawSample, now: Double) {
        ring.append(s)
        while let first = ring.first, now - first.t > RideConst.ringBufferS {
            ring.removeFirst()
        }
    }

    // MARK: - Triggers

    private func evaluateStart(sfMs2: Double, now: Double) {
        // rolling variance of specific force over the window
        varWindow.append((now, sfMs2))
        while let first = varWindow.first, now - first.t > RideConst.varWindowS {
            varWindow.removeFirst()
        }
        guard varWindow.count > 3 else { return }
        let mean = varWindow.reduce(0.0) { $0 + $1.v } / Double(varWindow.count)
        let variance = varWindow.reduce(0.0) { $0 + ($1.v - mean) * ($1.v - mean) } / Double(varWindow.count)

        if variance > RideConst.startVarThreshold {
            if highVarSince == nil { highVarSince = now }
            else if now - highVarSince! >= RideConst.startSustainS {
                beginRecording(seedFromRing: true)
            }
        } else {
            highVarSince = nil
        }
    }

    private func evaluateEnd(_ s: RawSample, now: Double) {
        if now - recordStart >= RideConst.maxDurationS {
            _ = finishRecording()
            return
        }
        if manual { return }  // manual recordings end only via stopRecording()

        // Quiet detection: variance of the last window below the end threshold.
        let recent = capture.filter { now - $0.t <= RideConst.varWindowS }
        guard recent.count > 3 else { return }
        let mean = recent.reduce(0.0) { $0 + $1.sfMs2 } / Double(recent.count)
        let variance = recent.reduce(0.0) { $0 + ($1.sfMs2 - mean) * ($1.sfMs2 - mean) } / Double(recent.count)
        if variance < RideConst.endVarThreshold {
            if lowVarSince == nil { lowVarSince = now }
            else if now - lowVarSince! >= RideConst.endSustainS {
                _ = finishRecording()
            }
        } else {
            lowVarSince = nil
        }
    }

    private func beginRecording(seedFromRing: Bool) {
        guard !recording else { return }
        recording = true
        lowVarSince = nil
        highVarSince = nil
        capture = seedFromRing ? ring : []
        recordStart = capture.first?.t ?? (ring.last?.t ?? Date().timeIntervalSince1970)
        // Altimeter already running from startMonitoring (W3) — do not restart it
        // here or the ring's pre-trigger altitude baseline resets to the trigger
        // point, losing the lift hill.
        startDeviceMotion(hz: RideConst.activeHz)  // escalate to full rate
        onRideStarted?()
    }

    private func finishRecording() -> RideResult? {
        guard recording else { return nil }
        recording = false
        let wasManual = manual
        manual = false
        let samples = capture
        capture = []
        // Leave the altimeter running — it's owned by the monitoring lifecycle
        // now (started in startMonitoring, stopped in stopMonitoring), so the
        // next ride's pre-trigger ring keeps getting real altitude (W3).

        // back down to monitoring rate (unless we're fully stopping)
        if monitoring { startDeviceMotion(hz: RideConst.monitorHz) }

        let result = RideMetricsComputer.compute(
            samples,
            baroAvailable: baroAvailable,
            gyroAvailable: gyroAvailable
        )
        if let result = result, !wasManual {
            onRideDetected?(result)
        }
        return result
    }
}
